### PR TITLE
[DPE-3700] Add check before scrape is configured

### LINKED ISSUE #194: OpenSearch fails if grafana-agent is related since the start
When deploying OpenSearch from `2/edge` and relating it right away with grafana-agent, it fails with:

```
2024-03-04 17:51:57,868 DEBUG    `cos-tool` unavailable. Leaving expression unchanged: sum by (cluster, instance, node) (opensearch_jvm_mem_heap_used_percent) > 75

2024-03-04 17:51:57,871 DEBUG    `cos-tool` unavailable. Leaving expression unchanged: sum by (cluster, instance, node) (opensearch_os_cpu_percent) > 90

2024-03-04 17:51:57,873 DEBUG    `cos-tool` unavailable. Leaving expression unchanged: sum by (cluster, instance, node) (opensearch_process_cpu_percent) > 90

2024-03-04 17:51:57,876 DEBUG    Reading <property object at 0x7f247ca65c10> rule from src/alert_rules/prometheus/prometheus_alerts.yaml
2024-03-04 17:51:57,879 DEBUG    Alert rules path does not exist: src/loki_alert_rules
2024-03-04 17:51:57,909 ERROR    Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-5/charm/./src/charm.py", line 94, in <module>
    main(OpenSearchOperatorCharm)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/venv/ops/main.py", line 436, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/venv/ops/framework.py", line 351, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/venv/ops/framework.py", line 853, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/venv/ops/framework.py", line 942, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-opensearch-5/charm/lib/charms/grafana_agent/v0/cos_agent.py", line 376, in _on_refresh
    metrics_scrape_jobs=self._scrape_jobs,
  File "/var/lib/juju/agents/unit-opensearch-5/charm/lib/charms/grafana_agent/v0/cos_agent.py", line 394, in _scrape_jobs
    scrape_configs = self._scrape_configs()
  File "/var/lib/juju/agents/unit-opensearch-5/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 1086, in _scrape_config
    ca = app_secrets.get("ca-cert")
AttributeError: 'NoneType' object has no attribute 'get'
```

Juju status:

```
Model       Controller         Cloud/Region   Version  SLA          Timestamp
opensearch  aws-tf-controller  aws/us-east-1  3.4.0    unsupported  18:59:14+01:00

SAAS                             Status  Store              URL
alertmanager-karma-dashboard     active  aws-tf-controller  admin/cos.alertmanager-karma-dashboard
grafana-grafana-dashboard        active  aws-tf-controller  admin/cos.grafana-grafana-dashboard
loki-logging                     active  aws-tf-controller  admin/cos.loki-logging
prometheus-metrics-endpoint      active  aws-tf-controller  admin/cos.prometheus-metrics-endpoint
prometheus-receive-remote-write  active  aws-tf-controller  admin/cos.prometheus-receive-remote-write

App                       Version  Status   Scale  Charm                     Channel        Rev  Exposed  Message
grafana-agent                      active       3  grafana-agent             stable          28  no       
opensearch                         active       3  opensearch                2/edge          39  yes      
self-signed-certificates           active       1  self-signed-certificates  latest/stable   72  no       
sysconfig                          blocked      3  sysconfig                 stable          33  no       update-grub and reboot required. Changes in: /etc/default/grub.d/90-sysconfig.cfg

Unit                         Workload     Agent      Machine  Public address   Ports  Message
opensearch/3                 maintenance  executing  4        192.168.235.136         Plugin configuration started.
  grafana-agent/2            active       idle                192.168.235.136         
  sysconfig/2                blocked      idle                192.168.235.136         update-grub and reboot required. Changes in: /etc/default/grub.d/90-sysconfig.cfg
opensearch/4                 maintenance  executing  5        192.168.235.232         Plugin configuration started.
  grafana-agent/1            active       idle                192.168.235.232         
  sysconfig/1                blocked      idle                192.168.235.232         update-grub and reboot required. Changes in: /etc/default/grub.d/90-sysconfig.cfg
opensearch/5*                active       executing  6        192.168.235.198         
  grafana-agent/0*           active       idle                192.168.235.198         
  sysconfig/0*               blocked      idle                192.168.235.198         update-grub and reboot required. Changes in: /etc/default/grub.d/90-sysconfig.cfg
self-signed-certificates/0*  active       idle       0        192.168.235.38          

Machine  State    Address          Inst id              Base          AZ          Message
0        started  192.168.235.38   i-0cbe3dd0e477c2b80  ubuntu@22.04  us-east-1a  running
4        started  192.168.235.136  i-09a8a15dbebd3cf1e  ubuntu@22.04  us-east-1a  running
5        started  192.168.235.232  i-0646338afc099528e  ubuntu@22.04  us-east-1a  running
6        started  192.168.235.198  i-0e7cd4c302b0d77bf  ubuntu@22.04  us-east-1a  running

Integration provider                                  Requirer                                     Interface                Type         Message
grafana-agent:grafana-dashboards-provider             grafana-grafana-dashboard:grafana-dashboard  grafana_dashboard        regular      
grafana-agent:peers                                   grafana-agent:peers                          grafana_agent_replica    peer         
loki-logging:logging                                  grafana-agent:logging-consumer               loki_push_api            regular      
opensearch:cos-agent                                  grafana-agent:cos-agent                      cos_agent                subordinate  
opensearch:juju-info                                  sysconfig:juju-info                          juju-info                subordinate  
opensearch:opensearch-peers                           opensearch:opensearch-peers                  opensearch_peers         peer         
opensearch:service                                    opensearch:service                           rolling_op               peer         
prometheus-receive-remote-write:receive-remote-write  grafana-agent:send-remote-write              prometheus_remote_write  regular      
self-signed-certificates:certificates                 opensearch:certificates                      tls-certificates         regular  
```

### PR DESCRIPTION
**Pull request by phvalguima**
Closes #194

### CODE REVIEW

**Review by juditnovak — APPROVED** [review]
Looks good w/o much insight on the context :-)

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 1]
```diff
@@ -1074,9 +1074,13 @@ def _check_certs_expiration(self, event: UpdateStatusEvent) -> None:
...
-        app_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
-        ca = app_secrets.get("ca-cert")
-        pwd = self.secrets.get(Scope.APP, self.secrets.password_key(COSUser))
+        if (
+            not (app_secrets := self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val))
+            or not (ca := app_secrets.get("ca-cert"))
+            or not (pwd := self.secrets.get(Scope.APP, self.secrets.password_key(COSUser)))
+        ):
+            # Not yet ready, waiting for certain values to be set
+            return []
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1083]: When is the right config set again?

Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1083]: @Mehdi-Bendriss this is related to the COS user setup. @juditnovak may correct me here but we need the start process to be over, as the user is added at post-start-init. By then, we should also receive the CA certs.

Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1083]: Yes - so my question is: once the CA is set, and COSUser created, how does the `self.cos_integration.scrape_configs` object get updated with these 2 new values (knowing it was already set to `[]`)?

Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1083]: @Mehdi-Bendriss that method is called every time the charm comes up, [here](https://github.com/canonical/opensearch-operator/blob/main/lib/charms/opensearch/v0/opensearch_base_charm.py#L148). So, it starts with "empty" but eventually, we can populate this method.

Btw, that is the standard way charms deal with COS provider, by calling them at the constructor of the main charm.

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**